query/sources/impl/src/test/groovy/org/codice/ddf/admin/sources/csw/discover/DiscoverCswActionTest.groovy

```diff
@@ -0,0 +1,260 @@
+
+    static URL_FIELD_PATH = [ADDRESS_FIELD_PATH, URL_NAME].flatten()
+
+    static CREDENTIALS_FIELD_PATH = [BASE_PATH, CREDENTIALS].flatten()
+
+    static USERNAME_FIELD_PATH = [CREDENTIALS_FIELD_PATH, USERNAME].flatten()
+
+    static PASSWORD_FIELD_PATH = [CREDENTIALS_FIELD_PATH, PASSWORD].flatten()
+
+    def setup() {
+        refreshDiscoverByAddressActionArgs()
+        refreshDiscoverByUrlActionArgs()
+        cswSourceUtils = Mock(CswSourceUtils)
+        discoverCsw = new DiscoverCswAction(cswSourceUtils)
+    }
+
+    def 'test successful discover using URL'() {
+        setup:
+        discoverCsw.setArguments(discoverByUrlActionArgs)
+
+        when:
+        def report = discoverCsw.process()
+
+        then:
+        1 * cswSourceUtils.getPreferredCswConfig(_ , _) >> createResult(false, SourceConfigUnionField.class)
+        report.result() instanceof SourceInfoField
+        ((SourceInfoField) report.result()).isAvailable()
+        ((SourceInfoField) report.result()).config() != null
+    }
+
+    def 'test successful discover using hostname and port'() {
+        setup:
+        discoverCsw.setArguments(discoverByAddressActionArgs)
+
+        when:
+        def report = discoverCsw.process()
+
+        then:
+        1 * cswSourceUtils.discoverCswUrl(_, _) >> createResult(false, UrlField.class)
+        1 * cswSourceUtils.getPreferredCswConfig(_, _) >> createResult(false, SourceConfigUnionField.class)
+        report.result() instanceof SourceInfoField
+        ((SourceInfoField) report.result()).isAvailable()
+        ((SourceInfoField) report.result()).config() != null
+    }
+
+    def 'test failure discovery using URL while getting preferred config'() {
+        setup:
+        discoverCsw.setArguments(discoverByUrlActionArgs)
+
+        when:
+        def report = discoverCsw.process()
+
+        then:
+        1 * cswSourceUtils.getPreferredCswConfig(_, _) >> createResult(true, null)
+        report.result() == null
+        report.messages().size() == 1
+    }
+
+    def 'test failure when using hostname+port when getting discovering the URL'() {
+        setup:
+        discoverCsw.setArguments(discoverByAddressActionArgs)
+
+        when:
+        def report = discoverCsw.process()
+
+        then:
+        1 * cswSourceUtils.discoverCswUrl(_, _) >> createResult(true, null)
+        report.result() == null
+        report.messages().size() == 1
+    }
+
+    def 'test failure discovery using hostname+port while getting preferred config'() {
+        setup:
+        discoverCsw.setArguments(discoverByAddressActionArgs)
+
+        when:
+        def report = discoverCsw.process()
+
+        then:
+        1 * cswSourceUtils.discoverCswUrl(_, _) >> createResult(false, UrlField.class)
+        1 * cswSourceUtils.getPreferredCswConfig(_, _) >> createResult(true, null)
+        report.result() == null
+        report.messages().size() == 1
+    }
+
+    def 'test missing required URL field when no arguments provided'() {
```

Comment:
We should collapse all these missing required field tests into one test. There is extra testing being done on the fields that should be moved to their corresponding test classes.